### PR TITLE
Common subexpression eliminator.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: test/libjulia/CommonSubexpression.cpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+		"{ let a := mload(1) let b := mload(1) }",
+		"{ let a := mload(1) let b := mload(1) }"
+	);
+}
+
+BOOST_AUTO_TEST_CASE(branches_if)
+{
+	CHECK(
+		"{ let b := 1 if b { b := 1 } let c := 1 }",
+		"{ let b := 1 if b { b := b } let c := 1 }"
```
Comment (axic) [inline test/libjulia/CommonSubexpression.cpp:85]: Will `b := b` removed by something? :)

Comment (chriseth) [reply test/libjulia/CommonSubexpression.cpp:85]: Not sure, we will have to see :)

**Comment by chriseth** [comment]
@ekpyron do you want to review this?

File: test/libjulia/CommonSubexpression.cpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	assembly::AsmPrinter p;\
+	Block b = disambiguate(_original, false);\
+	(CommonSubexpressionEliminator{})(b);\
+	string result = p(b);\
+	BOOST_CHECK_EQUAL(result, format(_expectation, false));\
+}\
+while(false)
+
+BOOST_AUTO_TEST_SUITE(IuliaCSE)
+
```
Comment (ekpyron) [inline test/libjulia/CommonSubexpression.cpp:51]: Shouldn't there be some negative examples involving for instance ``gasleft`` that will change between two invocations and can't be substituted? Is substitution prevented for these cases?

Comment (chriseth) [reply test/libjulia/CommonSubexpression.cpp:51]: There are examples in tests other optimizer stages which basically have the same requirement, but I can add some here, too.

Comment (ekpyron) [reply test/libjulia/CommonSubexpression.cpp:51]: How exactly are such cases prevented? To me it looks like ``SyntacticalEqualityChecker`` will consider function calls to the same function with the same arguments to be equal, even if the function may have side effects and calling once + reusing the return value is different from calling twice...

Comment (chriseth) [reply test/libjulia/CommonSubexpression.cpp:51]: Updated.

**Review by ekpyron — APPROVED** [review]

File: libjulia/optimiser/CommonSubexpressionEliminator.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * Optimisation stage that replaces expressions known to be the current value of a variable
+ * in scope by a reference to that variable.
+ */
+
+#pragma once
+
+#include <libjulia/optimiser/DataFlowAnalyzer.h>
+
+#include <string>
```
Comment (ekpyron) [inline libjulia/optimiser/CommonSubexpressionEliminator.h]: These includes don't seem to be necessary.

**Review by ekpyron — CHANGES_REQUESTED** [review]

**Review by ekpyron — APPROVED** [review]

**Pull request merged**